packages/gatsby/src/state-machines/data-layer/index.ts

```diff
@@ -0,0 +1,95 @@
+import { MachineConfig, Machine } from "xstate"
+import { dataLayerActions } from "./actions"
+import { IDataLayerContext } from "./types"
+import { dataLayerServices } from "./services"
+
+export type DataLayerResult = Pick<
+  IDataLayerContext,
+  | "gatsbyNodeGraphQLFunction"
+  | "graphqlRunner"
+  | "pagesToBuild"
+  | "pagesToDelete"
+>
+
+export const dataLayerStates: MachineConfig<IDataLayerContext, any, any> = {
+  initial: `customizingSchema`,
+  states: {
+    customizingSchema: {
+      invoke: {
+        src: `customizeSchema`,
+        id: `customizing-schema`,
+        onDone: {
+          target: `sourcingNodes`,
+        },
+      },
+    },
+    sourcingNodes: {
+      invoke: {
+        src: `sourceNodes`,
+        id: `sourcing-nodes`,
+        onDone: {
+          target: `buildingSchema`,
+          actions: `assignChangedPages`,
+        },
+      },
+    },
+    buildingSchema: {
+      invoke: {
+        id: `building-schema`,
+        src: `buildSchema`,
+        onDone: {
+          target: `creatingPages`,
+          actions: `assignGatsbyNodeGraphQL`,
+        },
+      },
+    },
+    creatingPages: {
+      invoke: {
+        id: `creating-pages`,
+        src: `createPages`,
+        onDone: [
+          {
+            target: `creatingPagesStatefully`,
+            actions: `assignChangedPages`,
+            cond: (context): boolean => !!context.firstRun,
+          },
+          {
+            target: `done`,
+            actions: `assignChangedPages`,
+          },
+        ],
+      },
+    },
+    creatingPagesStatefully: {
+      invoke: {
+        src: `createPagesStatefully`,
+        id: `creating-pages-statefully`,
+        onDone: {
+          target: `done`,
+        },
+      },
+    },
+    done: {
+      type: `final`,
+      data: ({
+        gatsbyNodeGraphQLFunction,
+        graphqlRunner,
+        pagesToBuild,
+        pagesToDelete,
+      }): DataLayerResult => {
+        return {
+          gatsbyNodeGraphQLFunction,
+          graphqlRunner,
+          pagesToBuild,
+          pagesToDelete,
+        }
+      },
+    },
+  },
+}
+
+// eslint-disable-next-line new-cap
```

Comment:
I imagine we will have to use this eslint directive anytime we create new `Machine` - maybe instead of those we should just add `Machine` to `new-cap` rule config exception via `capIsNewExceptions` option ( https://eslint.org/docs/rules/new-cap#capisnewexceptions )?

Reply:
(I'm bit allergic to those ad-hoc eslint-disable directives :P)

Reply:
Ah nice, I didn't realise you could add specific exceptions like that.

Reply:
https://github.com/gatsbyjs/gatsby/pull/25436